### PR TITLE
Create a KotlinStackFrameValueContributor Extension Point

### PR DESCRIPTION
**Pull request by alonalbert**
https://youtrack.jetbrains.com/issue/KTIJ-23529/Feature-Request-Add-an-Extension-Point-For-Contributing-Values-to-Debugger-View

### CODE REVIEW

File: java/java-impl/src/META-INF/JavaPlugin.xml [hunk 1]
```diff
@@ -237,6 +237,9 @@
     <extensionPoint qualifiedName="com.intellij.debugger.asyncStackTraceProvider"
                     interface="com.intellij.debugger.engine.AsyncStackTraceProvider" dynamic="true"/>
 
+    <extensionPoint qualifiedName="com.intellij.debugger.kotlinStackFrameValueContributor"
```
Comment (yanex) [inline java/java-impl/src/META-INF/JavaPlugin.xml:240]: Please register extension points in the Kotlin plugin if they're Kotlin-only.
This particular extension point should be registered in `plugins/kotlin/plugin/common/resources/META-INF/jvm-debugger.xml`.

File: plugins/kotlin/jvm-debugger/core/src/org/jetbrains/kotlin/idea/debugger/core/stackFrame/KotlinStackFrame.kt [hunk 2]
```diff
@@ -61,8 +60,12 @@ open class KotlinStackFrame(
         if (!removeSyntheticThisObject(evaluationContext, children, existingVariables) && thisVariables.isNotEmpty()) {
             remapThisObjectForOuterThis(evaluationContext, children, existingVariables)
         }
-
         children.add(evaluationContext, thisVariables, otherVariables)
+        KotlinStackFrameValueContributor.EP.extensions.forEach { contributor ->
+            contributor.contributeValues(this, evaluationContext, variables,).forEach {
```
Comment (yanex) [inline plugins/kotlin/jvm-debugger/core/src/org/jetbrains/kotlin/idea/debugger/core/stackFrame/KotlinStackFrame.kt:65]: Extra trailing comma.

File: plugins/kotlin/jvm-debugger/core/src/org/jetbrains/kotlin/idea/debugger/core/stackFrame/KotlinStackFrame.kt [hunk 3]
```diff
@@ -61,8 +60,12 @@ open class KotlinStackFrame(
         if (!removeSyntheticThisObject(evaluationContext, children, existingVariables) && thisVariables.isNotEmpty()) {
             remapThisObjectForOuterThis(evaluationContext, children, existingVariables)
         }
-
         children.add(evaluationContext, thisVariables, otherVariables)
+        KotlinStackFrameValueContributor.EP.extensions.forEach { contributor ->
```
Comment (yanex) [inline plugins/kotlin/jvm-debugger/core/src/org/jetbrains/kotlin/idea/debugger/core/stackFrame/KotlinStackFrame.kt:64]: We usually use the `for {}` loop instead (unless a callable reference is passed).

File: java/java-impl/src/META-INF/JavaPlugin.xml [hunk 4]
```diff
@@ -237,6 +237,9 @@
...
+                    interface="org.jetbrains.kotlin.idea.debugger.core.stackFrame.KotlinStackFrameValueContributor" dynamic="true"/>
```
Comment (yanex) [inline java/java-impl/src/META-INF/JavaPlugin.xml:241]: By the way, there is `com.intellij.debugger.engine.FrameExtraVariablesProvider`. Is it not enough for your use-cases?

**Review by yanex — CHANGES_REQUESTED** [review]

File: java/java-impl/src/META-INF/JavaPlugin.xml (hunk 1 above)
Comment (alonalbert) [reply java/java-impl/src/META-INF/JavaPlugin.xml:240]: Done

File: plugins/kotlin/jvm-debugger/core/src/org/jetbrains/kotlin/idea/debugger/core/stackFrame/KotlinStackFrame.kt (hunk 2 above)
Comment (alonalbert) [reply plugins/kotlin/jvm-debugger/core/src/org/jetbrains/kotlin/idea/debugger/core/stackFrame/KotlinStackFrame.kt:65]: Done

File: java/java-impl/src/META-INF/JavaPlugin.xml (hunk 4 above)
Comment (alonalbert) [reply java/java-impl/src/META-INF/JavaPlugin.xml:241]: I wasn't aware of this EP but unfortunately, it's not enough for my use case.

I think it should be possible to extend FrameExtraVariablesProvider in a way that will make it work.

I'm sending you an image of my use case in a private channel.

Comment (alonalbert) [reply java/java-impl/src/META-INF/JavaPlugin.xml:241]: My use requires adding a custom XValue to children. The existing EP doesn't allow that. It just seems allow adding a code fragment dressed as a variable.

As I said, I think I can extend it by adding another method to to that returns a list or XValue that can be added to children.

Do you want me to make this change or stick with my new EP?

**Review by yanex — APPROVED** [review]

**Comment by yanex** [comment]
Merged to `master` manually: https://github.com/JetBrains/intellij-community/commit/c0dadceeee702d07e00ecafedf1eb41021aa9f11.
Thanks!

**Pull request closed without merging**